ming/docs/man/gif2dbl.1
.TH gif2dbl 1 "25 Nov 2009" "" "Ming utils"
.\" $Id: gif2dbl.1,v 1.1 2009/11/26 03:28:52 anderson Exp $
.SH NAME
gif2dbl - convert a gif file to an dbl file using a RGB palette
.SH SYNOPSIS
.B gif2dbl <file.gif>
.SH DESCRIPTION
.B gif2dbl
is a command line tool for converting a gif file into a dbl file. Output is written to file.dbl.
H OPTIONS
.TP
\fB<file.gif>\fR
Name of gif format file to use as input.
.SH AUTHOR
Written by Raffaele Sena (raff@aromatic.org)
.SH COPYRIGHT
Code is released under the GNU GPL license.


libxl/changelog.txt
Version 3.4.2 (2013-04-25)
-------------------------------------------------

       - fixed NoteSh record (some xls files couldn't be loaded)

       - fixed RRTabId record
         (Protected view mode was activated in Excel 2010 and Excel 2013
         in some cases)

       - fixed Book::insertSheet() method
         (named ranges, print areas, repeat rows and columns didn't remain 
          in corresponding sheets)

       - fixed Style record (some xls files couldn't be loaded)

       - added quotes for names with hyphen (xlsx)

       - fixed bug in Sheet::delNamedRange() (xlsx)
       
       - fixed bug with saving some xls files with pictures
         (output file was corrupt)

       - fixed bug with crash in MinGW 4.7
         (added __cdecl keyword for any compilers)

       - fixed bug with Mac::Numbers compatibility

       - fixed bug with overlapping of calculated values 
         in cells with formula expression and format 
         for xlsx format (no screen refresh)

       - fixed bug in Sheet::writeFormula() method (xls)
         [ "incorrect token in formula" error message
         for some valid expressions ]

       - fixes for Apple clang compiler

       - updated ZipArchive library to version 4.5

       - fixed bug with column-based named ranges in
         Sheet::getNamedRange() and Sheet::namedRange() methods
         for xlsx format
         
       - fixed bug in Sheet::getPicture() for xlsx files

       - fixed bug in Sheet::insertRow() for xls files

       - fixed bug with loading some xlsx files ("mergeCell ref is invalid")

       - fixed bug with length of sheet name in UTF-8 encoding

       - fixed bug with loading some xlsx files ("bad end element" error)

       - fixed bug with SUMPRODUCT function in formula expression (xls)

       - fixed bug in Sheet::pictureSize() for xlsx files


Version 3.4.1 (2013-01-04)
--------------------------------------------------

       - added Book::sheetType() method for detection of chart sheet

       - fixed bug with loading some xls files

       - fixed bug in Sheet::delNamedRange() for xls files

       - fixed bug in Sheet::writeNum() method


Version 3.4.0 (2012-12-26)
--------------------------------------------------

       - added support of xlsm files

       - added UTF-8 support (mode is activated with book->setLocale("UTF-8"))

       - fixed bugs with loading some xls and xlsx files

       - fixed compatibility issues with Numbers for Mac

       - added support of empty cells in Sheet::cellFormat()

       - changed format picture property from "Move but don't size with cells" to "Move and size with cells"

       - fixed Sheet::setNamedRange(), Sheet::setPrintArea() and Sheet::setName() methods 
         for sheet's names with spaces (xlsx)

       - fixed bugs with some formula expressions with IF and AND functions (xls)
 
       - fixed bug with saving multi formatted cells in xlsx files

       - fixed bug in Book::delSheet() for xlsx files

       - fixed bug with reading some kind of error values

       - fixed issues with adding pictures into preloaded workbooks with pictures (xls)

       - fixed bug in Book::saveRaw() method (xls)

       - fixed possible memory leaks due to mismatched malloc/free functions (xlsx)

       - fixed bug in Sheet::cellFormat()

       - fixed bug with format/font inheritance for xlsx files

       - fixed Sheet::setPrintRepeatCols() method for xlsx files

       
Version 3.3.1 (2012-08-07)
--------------------------------------------------

       - fixed issue with adding pictures (red cross images) in xls files


Version 3.3.0 (2012-08-06)
--------------------------------------------------
Improvements:

       - added preserving macros and properties for xls files

       - added fullCalcOnLoad="true" attribute for xlsx files,
         instruction for Excel to calculate all formulas on loading

       - added defaultRowHeight property for xlsx files

       - changed internal version for xls files

       - added checking for corrupt xls files

Bug fixes:

       - fixed Sheet::getNamedRange(), Sheet::namedRangeSize() and Sheet::namedRange() methods (xls)

       - fixed Sheet::removeRow() for xls files

       - fixed bug in Sheet::getNamedRange() method (xlsx)

       - fixed bug with percent operator in formulas (xls)

       - fixed bug in Book::delSheet() method

       - fixed bug in Sheet::setPicture() method (xlsx)

       - fixed bug with comments (xls)


Version 3.2.4 (2012-04-25)
--------------------------------------------------
Improvements:

       - added two auxiliary methods for converting cell reference to row and column and vice versa

       - added Sheet::getTopLeftView() and Sheet::setTopLeftView() methods for getting/setting
         a view position in the sheet

       - added support for EMF pictures

       - added multisheet reference support in formulas in xls format, 
         for example: SUM(Sheet1:Sheet10!A1)

New methods:

       - Sheet::addrToRowCol()

       - Sheet::rowColToAddr()

       - Sheet::getTopLeftView()

       - Sheet::setTopLeftView()

Bug fixes:

       - fixed bug with loading some files

       - fixed performance issue with Sheet::setMerge() for xlsx files

       - fixed bug in OfficeArtBlip record

       - fixed bug with adding pictures to some xls files

       - fixed bug with macro records in xls files

       - fixed bug with writing strings with leading and trailing space symbols in xlsx format
         (added attribute xml:space="preserve" for such strings)

       - fixed bug in Sheet::setRowHidden() for xls format

       - fixed bug with getting named ranges in some xlsx files

       - fixed bug in Sheet::setCol() method for widths in range from 0 to 1 for xlsx format


Version 3.2.3 (2011-10-12)
--------------------------------------------------
Improvements:

       - added possibility to hide/unhide sheets
       
Bug fixes:


ming/util/gif2dbl.c
/* gif2dbl: convert a gif file to a DBL file using a RGB palette
 * for non-transparent gifs and RGBA palette for transparent ones.
 *
 * author: Raffaele Sena (raff@aromatic.org)
 *
 * (derived from gif2mask.c)
 */

/* $Id: gif2dbl.c,v 1.10 2007/07/15 19:40:03 krechert Exp $ */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

#include <ming_config.h>
#include <gif_lib.h>
#include <zlib.h>

void error(char *msg)
{
  printf("%s:\n\n", msg);
  PrintGifError();
  exit(-1);
}

/**
 * Get transparency color from graphic extension block
 *
 * Return: transparency color or -1
 */
int getTransparentColor(GifFileType * file)
{
  int i,returnvalue=-1;
  ExtensionBlock * ext = file->SavedImages[0].ExtensionBlocks;
 
  for (i=0; i < file->SavedImages[0].ExtensionBlockCount; i++, ext++) {

    if (ext->Function == GRAPHICS_EXT_FUNC_CODE) {
      if (ext->Bytes[0] & 1){   /* there is a transparent color */
        if (!ext->Bytes[3]) returnvalue=255; // exception
        else returnvalue=ext->Bytes[3]&0xff;
      }
    }

  }

  return returnvalue;
}

unsigned char *readGif(char *fileName, int *length, int *bytesPerColor)
{
  GifFileType *file;
  ColorMapObject *colorMap;
  unsigned char *bits;
  unsigned char *data;
  unsigned char *p;
  int i, nColors, size, alpha, bgColor, alignedWidth;

  if((file = DGifOpenFileName(fileName)) == NULL)
    error("Error opening file");

  if(DGifSlurp(file) != GIF_OK)
    error("Error slurping file");

  /* data should now be available */

	/* Local colormap has precedence over Global colormap */
  colorMap = file->Image.ColorMap ? file->Image.ColorMap : file->SColorMap;

  nColors = colorMap->ColorCount;
  alpha = getTransparentColor(file);

	/* bgColor is the background color to fill the bitmap with
	 * if an image is smaller than screen coordinates
	 */
  if (file->SColorMap)			/* There is a GlobalColorMap */
    bgColor = file->SBackGroundColor;   /* The BackGroundColor is meaningful */
  else
  if (alpha >= 0)			/* There is a transparency color */
    bgColor = alpha;			/* set the bgColor to tranparent */
  else
    bgColor = 0;			/* Don't know what to do here.
					 * If this doesn't work, we could
					 * create a new color and set the
					 * alpha-channel to transparent
					 * (unless we are using all the 256
					 * colors, in which case either we
					 * give up, or move to 16-bits palette
					 */

  *bytesPerColor = 3;
  if (alpha >= 0) *bytesPerColor += 1;	/* if tranparent color, use the alpha
					 * channel (RGBA instead of RGB)
					 */
	
	/* Ah! The Flash specs says scanlines must be DWORD ALIGNED!
	 * (but image width is the correct number of pixels)
	 */
  alignedWidth = (file->SWidth + 3) & ~3;

  /* size = size-of-header + size-of-colormap + size-of-bitmap */
  size = 6 + (nColors * *bytesPerColor) + (alignedWidth * file->SHeight);
  data = malloc(size);

  p = data;
  *p++ = 3; 				/* 8-bit image data */
  *p++ = (file->SWidth >> 0) & 0xFF;	/* image width */
  *p++ = (file->SWidth >> 8) & 0xFF;
  *p++ = (file->SHeight >> 0) & 0xFF;	/* image height */
  *p++ = (file->SHeight >> 8) & 0xFF;
  *p++ = nColors-1;			/* size(colorMap) - 1 */

	/* create ColorMap */
  for(i=0; i<nColors; ++i)
  {
    GifColorType c = colorMap->Colors[i];

    if (*bytesPerColor == 3) {	/* RGB */
      *p++ = c.Red;
      *p++ = c.Green;
      *p++ = c.Blue;
    } else {			/* RGBA */
#if 1				/* You would think that an alpha value of 0
				 * would mean fully transparent, but Flash
				 * player doesn't seem to think so.
				 * So, we just set the transparency color 
				 * as 0,0,0,0
				 */
    if (i != alpha) {
      *p++ = c.Red;
      *p++ = c.Green;
      *p++ = c.Blue;
      *p++ = 255;	/* Fully opaque */
    } else {
      *p++ = 0;		/* red */
      *p++ = 0;		/* green */
      *p++ = 0;		/* blue */
      *p++ = 0;		/* Fully transparent */
    }
#else
      *p++ = c.Red;
      *p++ = c.Green;
      *p++ = c.Blue;
      *p++ = (i != alpha) ? 255 : 0; /* set alpha to 0 for transparent color */
#endif
    }
  }

  bits = file->SavedImages[0].RasterBits;

  if (alignedWidth == file->SWidth
	&& file->Image.Width == file->SWidth 
	&& file->Image.Height == file->SHeight) {

	/* we are all nicely aligned and don't need to move the bitmap around.
	 * Just copy the bits into the output buffer.
	 */
    memcpy(p, bits, file->SWidth * file->SHeight);

  } else {
	/* here we need to pad the scanline or to move the bitmap around
	 * (if the image is not at 0,0 in the logical screen)
	 */
    int screenWidth = file->SWidth;
    int screenHeight = file->SHeight;

    int imageTop = file->Image.Top;
    int imageBottom = file->Image.Top + file->Image.Height;
    int imageLeft = file->Image.Left;
    int imageWidth = file->Image.Width;

    for (i=0; i < screenHeight; i++, p += alignedWidth) {

	/* the image is smaller than the logical "screen":
	 * Fill the reminder with the background color.
	 */
      if (imageWidth != screenWidth)
        memset(p, bgColor, screenWidth);

	/* finally, copy scanline
	 */
      if (i >= imageTop && i < imageBottom) {
	memcpy(p + imageLeft, bits, imageWidth);
	bits += imageWidth;
      }
    }
  }

	/* Done! */
  DGifCloseFile(file);

  *length = size;
  return data;
}

void usage()
{
  printf("gif2dbl - convert a gif image to DefineBitLossless2 bitmap\n");
  printf("\nusage: gif2dbl <file.gif>\n");

  exit(0);
}

int main(int argc, char *argv[])
{
  int len, size, bytesPerColor;
  unsigned long outsize;
  char *outfile;
  unsigned char *data, *outdata;
  FILE *dbl;

  if(argc <= 1)
    usage();

  len = strlen(argv[1]);

  if(strcmp(argv[1]+len-4, ".gif") != 0)
    usage();

  outfile = strdup(argv[1]);

  outfile[len-3] = 'd';
  outfile[len-2] = 'b';
  outfile[len-1] = 'l';

  data = readGif(argv[1], &size, &bytesPerColor);

  outdata = malloc(outsize = (int)floor(size*1.01+12));

  /* zlib-compress the gif data */
  compress2(outdata, &outsize, data+6, size-6, 9);

  /* dump to outfile */
  dbl = fopen(outfile, "wb");

  if (dbl)
  {
	/* DBL header */
    fputc('D', dbl);
    fputc('B', dbl);
    fputc('l', dbl);
    fputc(bytesPerColor==3 ? 1 : 2, dbl);	/* 1: RGB - 2: RGBA */

	/* fill size */
    fputc(((outsize+6) >> 24) & 0xFF, dbl);
    fputc(((outsize+6) >> 16) & 0xFF, dbl);
    fputc(((outsize+6) >> 8) & 0xFF, dbl);
    fputc(((outsize+6) >> 0) & 0xFF, dbl);

	/* write header */
    if(fwrite(data, sizeof(char), 6, dbl) != 6)
    {
      perror(outfile);
      exit(1);
    }
	/* write compressed data */
    if(fwrite(outdata, sizeof(char), outsize, dbl) != outsize)
    {
      perror(outfile);
      exit(-1);
    }

	/* and we are done! */
    fclose(dbl);
  }
  else {
    perror(outfile);
    exit(-1);
  }

  exit(0);
}
